pygments/lexers/mcfunction.py

```diff
@@ -0,0 +1,326 @@
+            (r'"', String.Double, "#pop"),
+        ],
+        "literals.string_single": [
+            (r"\\.", String.Escape),
+            (r"[^\\'\n]+", String.Single),
+            (r"'", String.Single, "#pop"),
+        ],
+
+        "compound": [
+            # this handles the unquoted snbt keys
+            #  note: stringified keys still work
+            (r"[A-Z_a-z]+", Name.Attribute),
+            include("operators"),
+            include("whitespace"),
+            include("literals"),
+            (r"\{", Punctuation.Compound.Push, "#push"),
+            (r"\[", Punctuation.List.New, "list"),
+            (r"\}", Punctuation.Compound.Pop, "#pop"),
+        ],
+
+        "list": [
+            #(r"[BIL](?:\;)", Keyword.Type),
+            (r"[A-Z_a-z]+", Name.Attribute),
+            include("literals"),
+            include("operators"),
+            include("whitespace"),
+            (r"\[", Punctuation.List.Push, "#push"),
+            (r"\{", Punctuation.Compound.New, "compound"),
+            (r"\]", Punctuation.List.Pop, "#pop"),
+        ],
+    }
+
+
+class MCFunctionLexer(RegexLexer):
+    """Lexer for the mcfunction scripting language used in Minecraft
+    
+
+    Modelled somewhat after the `GitHub mcfunction grammar <https://github.com/Arcensoth/language-mcfunction>`_.
+    .. versionadded:: 2.12.0
+    """
+
+    name = "MCFunction"
+    url = "https://minecraft.fandom.com/wiki/Commands"
+    aliases = ["mcfunction", "mcf"]
+    filenames = ["*.mcfunction"]
+    mimetypes = ["text/mcfunction"]
+
+    # Used to denotate the start of a block comment, borrowed from Github's mcfunction
+    _block_comment_prefix = "[>!*]"
+
+    tokens = {
+        "root": [
+            include("names"),
+            include("comments"),
+            include("literals"),
+            include("whitespace"),
+            include("property"),
+            include("operators"),
+            include("selectors"),
+        ],
+
+        "names": [
+            # The start of a command (either beginning of line OR after the run keyword)
+            #  We don't encode a list of keywords since mods, plugins, or even pre-processors
+            #  may add new commands, so we have a 'close-enough' regex which catches them.
+            (r"^(\s*)([a-z_]+)", bygroups(Whitespace, Name.Builtin)),
+            (r"(?<=run)\s+[a-z_]+", Name.Builtin),
+
+            # UUID
+            (
+                r"\b[0-9a-fA-F]+(?:-[0-9a-fA-F]+){4}\b",
+                Name.Variable,
+            ),
+            include("resource-name"),
+            # normal command names and scoreboards
+            #  there's no way to know the differences unfortuntely
+            (r"[A-Za-z_][A-Za-z0-9_.#%$]+", Keyword.Constant),
+            (r"[#%$][A-Za-z0-9_.#%$]+", Name.Variable.Magic),
+        ],
+
+        "resource-name": [
+            (
+                # resource names have to be lowercase
+                r"#?[a-z_][a-z_\.\-]*\:[a-z0-9_\.\-\/]+",
```

Comment:
No need to escape `.` and `/` in charsets, and `-` can be placed at the end. `:` is never special either.
```suggestion
                r"#?[a-z_][a-z_.-]*:[a-z0-9_./-]+",
```
Same below.